nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java

```diff
@@ -797,6 +797,184 @@ private void generateConstraintsForParam(
+   *     ones. This is an output parameter that gets mutated while generating the constraints to add
+   *     nested invocations.
+   * @param formalParamType The formal parameter type
+   * @param calledFromDataflow true if this inference is being done as part of dataflow analysis
+   * @param lambda The lambda argument
+   */
+  private void handleLambdaArgumentInGenericMethodInference(
+      VisitorState state,
+      @Nullable TreePath path,
+      ConstraintSolver solver,
+      Set<MethodInvocationTree> allInvocations,
+      Type formalParamType,
+      boolean calledFromDataflow,
+      LambdaExpressionTree lambda) {
+    Symbol.MethodSymbol fiMethod =
+        NullabilityUtil.getFunctionalInterfaceMethod(lambda, state.getTypes());
+
+    // get the return type of the functional interface method, viewed as a member of the formal
+    // parameter type, so the generic method's type variables are substituted in
+    Type.MethodType fiMethodTypeAsMember =
+        TypeSubstitutionUtils.memberType(state.getTypes(), formalParamType, fiMethod, config)
+            .asMethodType();
+    Type fiReturnType = fiMethodTypeAsMember.getReturnType();
+    Tree body = lambda.getBody();
+    if (body instanceof ExpressionTree) {
+      // Case 1: Expression body, e.g., () -> null
+      ExpressionTree returnedExpression = (ExpressionTree) body;
+      generateConstraintsForExprReturnedFromLambda(
+          returnedExpression,
+          state,
+          path,
+          solver,
+          allInvocations,
+          calledFromDataflow,
+          fiReturnType);
+    } else if (body instanceof BlockTree) {
+      // Case 2: Block body, e.g., () -> { return null; }
+
+      List<ExpressionTree> returnExpressions = ReturnFinder.findReturnExpressions(body);
+
+      for (ExpressionTree returnExpr : returnExpressions) {
+        generateConstraintsForExprReturnedFromLambda(
+            returnExpr, state, path, solver, allInvocations, calledFromDataflow, fiReturnType);
+      }
+    }
+  }
+
+  /**
+   * Generates type-inference constraints for an expression returned from a lambda argument, in the
+   * context of generic method inference. If the return expression is itself a generic method call,
+   * recursively generates constraints for that call.
+   *
+   * @param returnExpr the return expression to analyze
+   * @param state the visitor state
+   * @param path the tree path if available
+   * @param solver the constraint solver
+   * @param allInvocations a set of all method invocations that require inference, including nested
+   *     ones. This is an output parameter that gets mutated while generating the constraints to add
+   *     nested invocations.
+   * @param calledFromDataflow true if this inference is being done as part of dataflow analysis
+   * @param fiReturnType the return type of the functional interface method
+   */
+  private void generateConstraintsForExprReturnedFromLambda(
```

Comment:
Hrm, notice now how this logic is extremely similar to `generateConstraintsForParam`, but slightly different.  I feel like we could unify this logic...maybe have a common method like `generatePseudoAssignmentConstraint` that works for both parameter passing and returns.  And that would uniformly handle unwrapping parentheses, nested lambdas, dataflow, etc.  What do you think?